processing/src/main/java/org/apache/druid/segment/DimensionHandlerUtils.java

```diff
@@ -330,25 +329,67 @@ public static Long convertObjectToLong(@Nullable Object valObj, boolean reportPa
     } else if (valObj instanceof String) {
       Long ret = DimensionHandlerUtils.getExactLongFromDecimalString((String) valObj);
       if (reportParseExceptions && ret == null) {
-        throw new ParseException((String) valObj, "could not convert value [%s] to long", valObj);
+        final String message = (objectKey != null) ? StringUtils.nonStrictFormat(
+            "could not convert value [%s] to long for dimension [%s]",
+            valObj,
+            objectKey
+        ) : StringUtils.nonStrictFormat(
+            "could not convert value [%s] to long",
+            valObj
+        );
+        throw new ParseException((String) valObj, message);
       }
       return ret;
     } else if (valObj instanceof List) {
-      throw new ParseException(
-          valObj.getClass().toString(),
+      final String message = (objectKey != null) ? StringUtils.nonStrictFormat(
+          "Could not ingest value [%s] as long for dimension [%s]. A long column cannot have multiple values in the same row.",
+          valObj,
+          objectKey
+      ) : StringUtils.nonStrictFormat(
           "Could not ingest value %s as long. A long column cannot have multiple values in the same row.",
           valObj
       );
-    } else {
       throw new ParseException(
+          valObj.getClass().toString(),
+          message
+      );
+    } else {
+      final String message = (objectKey != null) ? StringUtils.nonStrictFormat(
+          "Could not convert value [%s] to long for dimension [%s]. Invalid type: [%s]",
+          valObj,
+          objectKey,
+          valObj.getClass()
+      ) : StringUtils.nonStrictFormat(
           valObj.getClass().toString(),
           "Could not convert value [%s] to long. Invalid type: [%s]",
           valObj,
           valObj.getClass()
       );
```

Comment:
nit: you can make this a bit concise and remove the multi-line ternary formatting for readability:
```suggestion
      final String message;
      if (objectKey != null) {
        message = StringUtils.nonStrictFormat(
            "Could not convert value [%s] to long for dimension [%s]. Invalid type: [%s]",
            valObj, objectKey, valObj.getClass()
        );
      } else {
        message = StringUtils.nonStrictFormat(
            "Could not convert value [%s] to long. Invalid type: [%s]", valObj, valObj.getClass()
        );
      }
```